src/parallel/parallel.hpp

```diff
@@ -122,8 +122,11 @@ class SegmentLock {
     unlock();
 
 private:
-    std::mutex mutex_;
-    // Using () causes the linter to complain, thinking counter_ is a function.
+    // Use a recursive mutex, which allows multiple lock operations by the same
+    // thread on the lock. However, if the number of locks exceeds an
+    // unspecified limit, then it will throw a std::system_error.
+    std::recursive_mutex mutex_;
+    // Using counter_(0) causes the linter to complain, thinking counter_ is a function.
     AtomicCounter counter_{0};
```

Comment:
You can also just use an assignment: `counter_ = 0`.

Reply:
This causes a linter error:

```
/src/parallel/parallel.hpp:130:30 [clang-diagnostic-error]
copying member subobject of type 'AtomicCounter' (aka 'atomic<unsigned int>') invokes deleted constructor
```

Reply:
Hmm the linter thinks we're doing a copy assignment but we're not. The 0 is an rvalue so it should be using overloads 1) and 2) from https://en.cppreference.com/w/cpp/atomic/atomic/operator%3D. It also compiles fine on my end.